app/src/main/java/com/example/android/codelabs/paging/ui/ReposAdapter.kt

```diff
@@ -21,37 +21,49 @@ import androidx.paging.PagingDataAdapter
 import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.RecyclerView.ViewHolder
 import com.example.android.codelabs.paging.R
-import com.example.android.codelabs.paging.model.Repo
 
 /**
  * Adapter for the list of repositories.
  */
-class ReposAdapter : PagingDataAdapter<Repo, ViewHolder>(REPO_COMPARATOR) {
+class ReposAdapter : PagingDataAdapter<UiModel, ViewHolder>(UIMODEL_COMPARATOR) {
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
-        return RepoViewHolder.create(parent)
+        return if (viewType == R.layout.repo_view_item) {
+            RepoViewHolder.create(parent)
+        } else {
+            SeparatorViewHolder.create(parent)
+        }
     }
 
     override fun getItemViewType(position: Int): Int {
-        return R.layout.repo_view_item
+        return when (getItem(position)) {
+            is UiModel.RepoItem -> R.layout.repo_view_item
+            is UiModel.SeparatorItem -> R.layout.separator_view_item
+            null -> 0
+        }
     }
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) {
-        val repo = getItem(position)
-        (holder as RepoViewHolder).bind(repo)
+        val uiModel = getItem(position)
+        uiModel.let {
+            when (uiModel) {
+                is UiModel.RepoItem -> (holder as RepoViewHolder).bind(uiModel.repo)
+                is UiModel.SeparatorItem -> (holder as SeparatorViewHolder).bind(uiModel.description)
+            }
+        }
     }
 
     companion object {
-        private val REPO_COMPARATOR = object : DiffUtil.ItemCallback<Repo>() {
-            override fun areItemsTheSame(oldItem: Repo, newItem: Repo): Boolean {
-                return if (oldItem is Repo && newItem is Repo) {
-                    oldItem == newItem
+        private val UIMODEL_COMPARATOR = object : DiffUtil.ItemCallback<UiModel>() {
```

Comment:
The code as written will cause all separators to shimmer, it should be checking for separators too

Should be fine to say all separators are the same, since DiffUtil will be doing greedy matching

I don't think any of this matters in behavior here, since without swipe to refresh, you're not actually going to see any diffing :man_shrugging:  but still good to recommend correct code in case they do

Reply:
added separator as well